plugin.rb

```diff
@@ -46,4 +47,77 @@ class Engine < ::Rails::Engine
       DiscourseLivestream.handle_topic_chat_channel_creation(topic)
     end
   end
+
+  register_modifier(:channel_memberships) do |f, user|
```

Comment:
I'm not sure I understand why we're doing it this way. I would have thought that the memberships would be created/updated when users join/leave the event associated with the "livestream". The users/groups allowed to join the associated chat would be controlled by the groups allowed to join the event.

Reply:
The main problem here is the chat's behavior. Currently, users can't preview channels they can't follow. Initially, I implemented this in the guardian extension for chat but those users who couldn't follow couldn't preview the chat so it defeats the purpose of the plugin. This modifiers aim at changing the state of the membership depending on user permissions so we dont have to modify core, which was recentlyish modified to behave this way

Reply:
> Currently, users can't preview channels they can't follow.

That's not true. Users can preview all the channels they have access to.

Reply:
It is my understanding this commit changed that, this is in the `channels_controller` https://github.com/discourse/discourse/commit/436b68a5819e01b8f08f1d645c8987505e109c4a. It changed the guardian use of `can_preview_chat_channel?` to `can_join_chat_channel?`. 

https://github.com/discourse/discourse/blob/1ae35ecce8e70da065906a56171dbaa171bf2bfe/plugins/chat/app/controllers/chat/api/channels_controller.rb#L115

Reply:
Yes, that's what I meant. If they can join a channel, they can preview it.

Reply:
Oh yeah but that is the issue since users won't know of the existence of the chat until they are given permission to join

Reply:
And they should be given the permission to join when they click "going" to the event of the livestream, shouldn't they?

Reply:
Currently that button only triggers the create in `ChannelsCurrentUserMembershipController` but then again that one checks if the user can join using the same guardian method. Who can join a livestream channle or not is now determined by the site setting `livestream_chat_allowed_groups`

Reply:
Where is the code that is triggered when a user clicks on "join" on an event?

Reply:
Oh I forgot its not the `ChannelsCurrentUserMembershipController` but the `ChannelManager` that gets called which is why I had to intervene the code there (If the user can't join due to the settings don't allow him to follow and if settings change the status should reflect accordingly)
https://github.com/discourse/discourse-livestream/blob/52af31013c5473947397b5b220634c28065a89e6/assets/javascripts/discourse/initializers/chat-overrides.js#L17

Reply:
That doesn't answer my question.

Reply:
The snipped of code I linked belongs to the function triggered when an user joins an event

Reply:
This should be done on the server.

Reply:
I think we can fix that, it would require changes in the calendar plugin as well since it lacks the event trigger.

Reply:
I think we can fix that, it would require changes in the calendar plugin as well since it lacks the event trigger.